### ISSUE TITLE
Append config values without overwriting existing ones

### ISSUE THREAD

**Issue by arkena00** [xmake-io/xmake#2401]
### Is your feature request related to a problem? Please describe.

If we just want to append a value to the config, we have to rewrite all the values

`xmake f --option1=value`

If we just want to add a value
`xmake f --option2=value`

The option1 is cleared

### Describe the solution you'd like

Let the choice to overwrite or append values

### Describe alternatives you've considered

- A flag to overwrite or append values
`xmake f -w --option1=value` or `xmake f -a --option1=value`

- A user confirmation to append or overwrite
`xmake f --option1=value`
`$ Overwrite (o) or append (a) config values ?`

- Clear values explicitly
`xmake f --option1= --option2=value`

### Additional context

Maybe `xmake f` should display current values to help the user seeing the state of his configuration

_No response_

**Comment by SirLynix** [comment]
That's a feature request I was thinking of. Maybe a `xmake patch` or `xmake f -p` which would keep current options values and only change others.

**Comment by waruqi** [comment]
Yes, every configuration is a complete configuration. Just like autoconf/cmake.

xmake also does not support incremental configuration, it is too complex to implement and requires dealing with dependencies between configuration parameters.

Even with the current handling of configuration, which is already very complex, I wouldn't consider introducing incremental configuration as a feature.

**Comment by SirLynix** [comment]
I know we've discussed this already but I really don't understand why it would be more complex than what we have today.

When configuring xmake for my project, I have to do this:
`xmake f --mode=debug ffmpeg=y usepch=y shadernodes=y tests=y`

More options are coming.

So when I want to change the mode/arch/plat, or even just disable the ffmpeg plugin, I have to retype the whole command:
`xmake f --mode=releasedbg ffmpeg=y usepch=y shadernodes=y tests=y`

A simple solution here would be to support a patch parameter, that keeps current values **as if** I retyped them.

For example

Initial setup:
`xmake f --mode=debug ffmpeg=y usepch=y shadernodes=y tests=y`

Then I do:
`xmake f -p --mode=releasedbg`

This would be exactly as if I typed this:
`xmake f ffmpeg=y usepch=y shadernodes=y tests=y --mode=releasedbg`

Options not specified just keep their original value, as if I wrote them, and new values are appended next.

There is no difference between xmake rewriting configs in the command-line it already known and me retyping them, except it's very cumbersome to do with a lot of options.

There's no added complexity for xmake, that's even what it does when using the GUI

This is already what xmake does with its GUI:
![](https://user-images.githubusercontent.com/3002461/170927519-3978bac8-a368-4c3e-9fcb-6cfa7ef46606.png)

GUI is fine but isn't as fast as typing commands (and doesn't work on CI)

**Comment by waruqi** [comment]
Many parameter options have dependencies, and may even have cascading dependencies.

You change a parameter value and all other parameter values that were previously cached, need to have them automatically updated based on that value. Otherwise the whole configuration may be invalid and the compilation will not work properly.

Initial setup:

```bash
$ xmake f -a x86_64
```

Then I do:

```bash
$ xmake f -p android
```

This would be exactly as if I typed this? 

```bash
$ xmake f -p android -a x86_64
```

oh, it's incorrect.  The cached `--arch` value should be changed to `arm64-v8a` instead of `x86_64`, otherwise it will build fails.

And this is just the simplest of dependencies, we may need to analyse all the parameter values being cached, which ones need to be changed and what values they need to be automatically updated to.

And then whether the parameter values of its two levels of dependencies should also need to be automatically synchronised.

Also, if there are user-defined options, how should they be handled if there are dependencies between them. And so on.

**Comment by SirLynix** [comment]
I get your point. It seems to me the heart of the problem is that `arch` / `plat` / `mode` shouldn't be at the same level as a project options, and that we should be able to change arch/plat/mode without touching a project option.

**Comment by waruqi** [comment]
I'm not just talking about plat/arch/mode, I'm just using them as an example. In the real world, any parameter is likely to have the problems I mentioned, including the user-defined option 

If you do not deal with dependencies, then the configuration state is completely dependent on the user's input and the probability of the configuration becoming invalid is very high.

However not all users are good at resolving them and they get frustrated by the compilation problems caused by the configuration and go spend more time trying to resolve them.

I am also receiving more issues to help users resolve compilation issues caused by various cache configurations.

Translated with www.DeepL.com/Translator (free version)

**Comment by waruqi** [comment]
Moreover, the logic for handling full configuration is now so complex that I have to be very careful every time I change this piece of code, otherwise it could easily introduce other unknown bugs.

https://github.com/xmake-io/xmake/blob/master/xmake/actions/config/main.lua

So I really don't want to add to its complexity, which would add a huge amount of extra complexity and create more problems, not only for the developers but also for the users.

**Comment by SirLynix** [comment]
This is why I think the default behavior should stay the same, and incremental should be explicitly requested (using --patch arg for example).

There would be no difference between typing
```
xmake f --zlib=y --png=y --openssl=y
xmake f --patch --smallest=y
```
and
```
xmake f --zlib=y --png=y --openssl=y --smallest=y
```

both leads to errors, due to user usage.

I really get your point and why you're against this feature, however this wouldn't be a problem for most developpers. The fact is that options are almost unusable when you have a lot of them, because you need to write them everything you want to change a small thing, like the mode (which do change very frequently).

In fact, I've been thinking of just writing my options to a .lua file included in my xmake.lua instead of using xmake options because of this, everytime I add an option it becomes more and more problematic.

I understand the problems that can arise with such a proposal, but my point is that those problems are already there since we cannot prevent users from setting incorrects options even today.

Something like 
```$ xmake f --patch --smallest=y
new options: 
 - smallest=y 
 - zlib=y 
 - png=y 
 - openssl=y

*error messages
```

would be enough for a developper to see where the problem is.

> Moreover, the logic for handling full configuration is now so complex that I have to be very careful every time I change this piece of code, otherwise it could easily introduce other unknown bugs.

I understand, but a -p parameter like I suggest wouldn't require you to even touch existing code.

To me, it seems we're trying to prevent issues for 1% projects (which can be fixed by simply not using `-p` in xmake config) where it could be useful for most devs using xmake.

**Comment by waruqi** [comment]
In fact, I've been thinking of just writing my options to a .lua file included in my xmake.lua instead of using xmake options because of this, everytime I add an option it becomes more and more problematic.

you can use `xmake f --import=/xxx/myconfigs.txt` and `xmake f --export`

> I understand the problems that can arise with such a proposal, but my point is that those problems are already there since we cannot prevent users from setting incorrects options even today.

But the full configuration will allow the user to reduce the probability of error as much as possible.

`xmake f --smallest=y` and `xmake f --zlib=y --png=y --openssl=y` 
Whichever configuration the user enters is correct.

whereas if the user enters the wrong `xmake f --zlib=y --png=y --openssl=y --smallest=y` in its entirety,
basically, the user can quickly tell for themselves that this is a conflicting configuration.

The following cache configuration, however, would be so obscure that the user would not be able to find the cause of the error in the first place. They would need to run `xmake f -v` to check the current configuration items one by one and then spend a lot of time to fix the configuration problem. 

```
$ xmake f --zlib=y --png=y --openssl=y
$ xmake f --smallest=y
```

**Comment by SirLynix** [comment]
I just tested `xmake f --export=config.txt` and `xmake f --import=config.txt --mode=releasedbg`, it seems to me it presents exactly the same dangers as xmake config-update, as it allows one to take existing config and override only some values. It presents exactly the same dangers as a `xmake config-update` command, yet is very useful.

**Comment by SirLynix** [comment]
I implemented it as a xmake plugin, if anybody wants:

xmake.lua:
```
function _project_menu_options()
    import("core.project.menu")
    return menu.options()
end

-- define task
task("config-update")

    -- set category
    set_category("plugin")

    -- on run
    on_run("main")

    -- set menu
    set_menu {
                -- usage
                usage = "xmake config-update [options]"

                -- description
            ,   description = "Incremental config update."

                -- options
            ,   options = {
                    {category = "."},
                    {'p', "plat",       "kv", "auto"    , "Compile for the given platform.", values = _plat_values},
                    {'a', "arch",       "kv", "auto"    ,   "Compile for the given architecture.", _arch_description, values = _arch_values},
                    {'m', "mode",       "kv", "auto" ,      "Compile for the given mode.", values = _mode_values},
                    {'k', "kind",       "kv", "static"  ,   "Compile for the given target kind.", values = {"static", "shared", "binary"}},
                    {nil, "host",       "kv", "$(host)" ,   "Set the current host environment."},
                    {nil, "policies",    "kv", nil       ,  "Set the project policies.",
                                                            "    e.g.",
                                                            "    - xmake f --policies=package.fetch_only",
                                                            "    - xmake f --policies=package.precompiled:n,package.install_only"},
                    {category = "Package Configuration"},
                    {nil, "require",    "kv",   nil     ,   "Require all dependent packages?", values = {"yes", "no"}},
                    {nil, "pkg_searchdirs", "kv", nil       , "The search directories of the remote package."
                                                            , "    e.g."
                                                            , "    - xmake f --pkg_searchdirs=/dir1" .. path.envsep() .. "/dir2"},
                    _project_menu_options,
        }
    }
```

main.lua:
```
import("core.base.option")
import("core.base.task")
import("core.cache.localcache")
import("core.project.project")

-- filter option
function _option_filter(name)
    local options =
    {
        target      = true
    ,   file        = true
    ,   root        = true
    ,   yes         = true
    ,   quiet       = true
    ,   confirm     = true
    ,   project     = true
    ,   verbose     = true
    ,   diagnosis   = true
    ,   require     = true
    ,   export      = true
    ,   import      = true
    }
    return not options[name]
end

function main()
    -- lock the whole project
    project.lock()

    local options = localcache.get("config", "options")

    for name, value in pairs(option.options()) do
        if _option_filter(name) then
            options = options or {}
            options[name] = value
        end
    end

    -- ensure options are well checked
    options["clean"] = true

    task.run("config", options)

    -- unlock the whole project
    project.unlock()
end
```

It probably can be improved, but it does what I need
```
C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config -c
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022

C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config-update --tests=y
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022

C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config-update --shadernodes=y
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022

C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config-update --usepch=y
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022

C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config -v
configure
{
    tests = true
    usepch = true
    shadernodes = true
...
}

C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config-update --shadernodes=n
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022

C:\Projets\Perso\NazaraNext\NazaraEngine>xmake config -v
configure
{
    shadernodes = false
    usepch = true
    tests = true
}
```

it's probably not perfect but it works quite well for my usage.

**Comment by arkena00** [comment]
Yep, i still don't understand, we all have to use our homemade solutions. Nobody would rewrite all his config values by hand.
Saving config values is just at user input level so how it could be a problem for xmake ?

**Comment by cpcgskill** [comment]
Is it possible to choose an option to read the configuration from a file? For example:
```python
{
    "mode": "debug",
    "vs": "2017",
}
```

**Comment by waruqi** [comment]
> Is it possible to choose an option to read the configuration from a file? For example:
> 
> ```python
> {
>     "mode": "debug",
>     "vs": "2017",
> }
> ```

xmake f --import/--export

**Comment by cpcgskill** [comment]
> > 是否可以选择一个选项来从文件中读取配置？例如：
> > ```python
> > {
> >     "mode": "debug",
> >     "vs": "2017",
> > }
> > ```
> 
> xmake f --导入/--导出

Thank you,ヾ(≧▽≦*)o
